packages/flutter/lib/src/cupertino/loupe.dart

```diff
@@ -0,0 +1,266 @@
+  double _verticalFocalPointAdjustment = 0;
+  late AnimationController _ioAnimationController;
+  late Animation<double> _ioAnimation;
+
+  @override
+  void initState() {
+    _ioAnimationController = AnimationController(
+      value: 0,
+      vsync: this,
+      duration: CupertinoLoupe._kIoAnimationDuration,
+    )..addListener(() => setState(() {}));
+
+    widget.controller.animationController = _ioAnimationController;
+    widget.loupeSelectionOverlayInfoBearer
+        .addListener(_determineLoupePositionAndFocalPoint);
+
+    _ioAnimation = Tween<double>(
+      begin: 0.0,
+      end: 1.0,
+    ).animate(
+        CurvedAnimation(parent: _ioAnimationController, curve: Curves.easeOut));
+
+    super.initState();
+  }
+
+  @override
+  void dispose() {
+    widget.controller.animationController = null;
+    widget.loupeSelectionOverlayInfoBearer
+        .removeListener(_determineLoupePositionAndFocalPoint);
+    super.dispose();
+  }
+
+  @override
+  void didChangeDependencies() {
+    _determineLoupePositionAndFocalPoint();
+    super.didChangeDependencies();
+  }
+
+  void _determineLoupePositionAndFocalPoint() {
+    final LoupeSelectionOverlayInfoBearer textEditingContext =
+        widget.loupeSelectionOverlayInfoBearer.value;
+
+    // The exact Y of the center of the current line.
+    final double verticalCenterOfCurrentLine =
+        textEditingContext.handleRect.center.dy;
+
+    // If the loupe is currently showing, but we have dragged out of threshold,
+    // we should hide it.
+    if (verticalCenterOfCurrentLine -
+            textEditingContext.globalGesturePosition.dy <
+        CupertinoTextEditingLoupe._kHideIfBelowThreshold) {
+      // Only signal a hide if we are currently showing.
+      if (widget.controller.shown) {
+        widget.controller.hide(removeFromOverlay: false);
+      }
+      return;
+    }
+
+    // If we are gone, but got to this point, we shouldn't be: show.
+    if (!widget.controller.shown) {
+      _ioAnimationController.forward();
+    }
+
+    // Never go above the center of the line, but have some resistance
+    // going downward if the drag goes too far.
+    final double verticalPositionOfLens = math.max(
+        verticalCenterOfCurrentLine,
+        verticalCenterOfCurrentLine -
+            (verticalCenterOfCurrentLine -
+                    textEditingContext.globalGesturePosition.dy) /
+                CupertinoTextEditingLoupe._kDragResistance);
+
+    // The raw position, tracking the gesture directly.
+    final Offset rawLoupePosition = Offset(
+        textEditingContext.globalGesturePosition.dx -
+            CupertinoLoupe.kSize.width / 2,
+        verticalPositionOfLens -
+            (CupertinoLoupe.kSize.height -
+                CupertinoLoupe.kVerticalFocalPointOffset));
+
+    final Rect screenRect = Offset.zero & MediaQuery.of(context).size;
+
+    // Adjust the loupe position so that it never exists outside the horizontal
+    // padding.
```

Comment:
While technically it is being adjusted on both Y and X, the Y constraints are loosened so that the entire loupe always may exist out of bounds. This is because on iOS the loupe doesn't respect any vertical bounds, it can just go right off the screen verticaly willy nilly